cpp/src/dataframe/cudf_table.cuh

```diff
@@ -472,6 +472,34 @@ public:
 
   };
 
+  __device__
+  void* get_column_device_pointer(size_type column_index)
+  {
+    return d_columns_data[column_index];
+  }
+
+  __device__
+  gdf_error get_row_valids(size_type row_index, gdf_valid_type * row_valid_byte_buffer) const
+  {
+    if(nullptr == row_valid_byte_buffer) {
+      return GDF_DATASET_EMPTY;
+    }
+    
+    for(size_type i = 0; i < num_columns; i++)
+    {
+      // get validity of item in column in self
+      byte_type validity = gdf_is_valid(d_columns_valids[i], row_index);
+      // set validity in output buffer
+      row_valid_byte_buffer[i / GDF_VALID_BITSIZE] |= (validity << (i % GDF_VALID_BITSIZE));
```

Comment:
> This update needs to be done atomically.

Actually, not really. The write location is exclusive to one thread as each thread takes care of one write buffer (output column). If two threads access the same word location, it's not how this function was supposed to be used and the entire output is invalid.

Reply:
> I think this whole function can be made a lot more efficient using warp-level intrinsics with __ballot_sync.

This won't be useful here because of the different access pattern
**Yours** - Threads writing to same buffer
```
source gather_map destination threads
 ┌─┐     ┬─┐        ┌─┐        ┌─┐
 ├─┤╶──→ ├─┤        ├─┤        ├─┤
 ├─┤ ┌╴→ ├─┤        ├─┤        ├─┤
 ├─┤╶┘┌→ └─┘        └─┘        └─┘
 ├─┤  │
 ├─┤╶─┘
 └─┘
```
**Mine** - threads writing to separate buffers
```
source destination thread
┬─┬┌─┐  ┬──┬──┐    ┌─┐
│ ││ │→ └──┴──┘    └─┘
├─┤├─┤  ┬──┬──┐    ┌─┐
│ ││ │→ └──┴──┘    └─┘
├─┤├─┤  ┬──┬──┐    ┬─┐
│ ││ │→ └──┴──┘    └─┘
└─┘└─┘
```
I know it's inefficient but I did this because we deal with long and skinny matrices and so to maximize the number of threads, I decided to make  individual threads own output columns rather than input columns. Either way, we make one coalesced and one uncoalesced memory access. I did initially plan to use `cub::BlockExchange` to make both accesses coalesced but decided against it because it would be premature optimization.

Reply:
Seems like a warp per output column would make more sense. Because that way you have no divergence on column, and you can write up to 32 output rows (equivalent to 32 input columns) per warp efficiently in a coalesced way.